File: spt3g_software/mapmaker/python/aberrationmodules.py
from spt3g.mapmaker import RelativisticAberrationPointing
from spt3g import core
from spt3g.core.util import Rename, Delete

@core.pipesegment_nodoc
def aberrate_alpha_delta(pipe,
        peculiar_ra,
        peculiar_dec,
        peculiar_vel,
        coord_sys,
        alpha_key,
        delta_key):
    """
    Modify the alpha-delta pointings in each scan frame to account for the
    effects of relativistic aberration in the given frame of reference.

    peculiar_ra: Right ascension of the observer's peculiar velocity
    peculiar_dec: Declination of the observer's peculiar velocity
    peculiar_vel: Magnitude of observer's peculiar velocity
    coord_sys: Coordinate system corresponding to the scan
    detector_alpha_key:  Where the detector alpha pointings is stored
    detector_delta_key: Where the detector delta pointing is stored
    """

    tmp_alpha_key="tmp_"+alpha_key
    tmp_delta_key="tmp_"+delta_key

    pipe.Add(RelativisticAberrationPointing,
            peculiar_ra=peculiar_ra,
            peculiar_dec=peculiar_dec,
            peculiar_vel=peculiar_vel,
            coord_sys=coord_sys,
            detector_alpha_key=alpha_key,
            detector_delta_key=delta_key,
            detector_alpha_out_key=tmp_alpha_key,
            detector_delta_out_key=tmp_delta_key)
    pipe.Add(Delete, keys=[alpha_key, delta_key], type=core.G3FrameType.Scan)
    pipe.Add(Rename, keys={tmp_alpha_key:alpha_key, tmp_delta_key:delta_key},
            type=core.G3FrameType.Scan)

File: spt3g_software/mapmaker/src/aberrationmodules.cpp
#include <pybindings.h>
#include <string>

#include <mapmaker/aberrationmodules.h>
#include <mapmaker/aberrationutils.h>

void RelativisticAberrationPointing::Process(G3FramePtr frame, std::deque<G3FramePtr> &out){
    if (frame->type == G3Frame::Scan){
        auto alphas = frame->Get<G3MapVectorDouble>(detector_alpha_key_);
        auto deltas = frame->Get<G3MapVectorDouble>(detector_delta_key_);

		G3MapVectorDoublePtr alpha_out(new G3MapVectorDouble());
		G3MapVectorDoublePtr delta_out(new G3MapVectorDouble());

		aberrate_delta_alpha(*alphas, *deltas, coordinate_system_, *alpha_out,
                   *delta_out, peculiar_vel_, peculiar_ra_/G3Units::rad,
                   peculiar_dec_/G3Units::rad);
		frame->Put(detector_alpha_out_key_, alpha_out);
		frame->Put(detector_delta_out_key_, delta_out);
    }
    out.push_back(frame);
}

EXPORT_G3MODULE("mapmaker", RelativisticAberrationPointing,
        (init<double,double,double,MapCoordReference,std::string,std::string,
         std::string,std::string>(
            ( arg("peculiar_ra")="PeculiarRA",
              arg("peculiar_dec")="PeculiarDec",
              arg("peculiar_vel")="PeculiarVel",
              arg("coord_sys")="CoordinateSys",
			  arg("detector_alpha_key")="DetectorAlphaPointing",
			  arg("detector_delta_key")="DetectorDeltaPointing",
			  arg("detector_alpha_out_key")="OutputAlphaPointing",
			  arg("detector_delta_out_key")="OutputDeltaPointing"))),
        RELATIVISTIC_ABERRATION_POINTING_DOCSTR)
